### PR TITLE
MPItrampoline: New package

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

File: M/MPItrampoline/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+cmake --build . --config RelWithDebInfo --parallel $nproc --target install
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms()
+# Apple: Dynamically loaded libraries (`dlopen`) are not handled correctly.
+# Windows: Does not have `dlopen`.
+# musl: Does not define `RTLD_DEEPBIND` for `dlopen`.
+platforms = filter(p -> !(Sys.isapple(p) || Sys.iswindows(p) || libc(p) == "musl")), platforms)
```
Comment (giordano) [inline M/MPItrampoline/build_tarballs.jl:36]:
```suggestion
platforms = filter(p -> !(Sys.isapple(p) || Sys.iswindows(p) || libc(p) == "musl"), platforms)
```

**Comment by giordano** [comment]
This is exciting!

File: M/MPItrampoline/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    LibraryProduct("libmpitrampoline", :libmpitrampoline),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae")),
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```
Comment (giordano) [inline M/MPItrampoline/build_tarballs.jl:61]: Can we require Julia v1.6, just to be future-proof?
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6")
```

Comment (eschnett) [reply M/MPItrampoline/build_tarballs.jl:61]: Sure! What does this do?

Comment (giordano) [reply M/MPItrampoline/build_tarballs.jl:61]: The generated JLL package will require Julia v1.6.  The main advantage of Julia v1.6 is that it doesn't freak out if there are unknown platforms in `Artifacts.toml`, which lets us add more easily new platforms in the future.  For example, `aarch64-apple-darwin` (Apple Silicon) isn't known to Julia v1.5-, and it doesn't like this if this platform shows up in `Artifacts.toml`

Comment (eschnett) [reply M/MPItrampoline/build_tarballs.jl]: Ok. Done.

**Comment by eschnett** [comment]
I want to test whether other Yggdrasil packages can be built against `MPItrampoline` before merging it. I've built and deployed `MPItrampoline` locally, but I can't seem to use it to build e.g. `ADIOS2`:
```
$ julia --color=yes build_tarballs.jl --verbose --debug --deploy=local x86_64-apple-darwin-libgfortran4
[ Info: Building and deploying version 2.7.1+7 to /Users/eschnett/.julia/dev/ADIOS2_jll
[ Info: Building for x86_64-apple-darwin-libgfortran4
┌ Warning: Unable to resolve MPItrampoline_jll
└ @ BinaryBuilderBase ~/.julia/packages/BinaryBuilderBase/q7lkZ/src/Dependencies.jl:193
ERROR: LoadError: Invalid dependency specifications!
```
The error comes from a line in `Dependencies.jl` where an expression `dep.uuid` is checked. Apparently, the uuid is `nothing`, and that's not allowed.

Does this mean that there needs to be a public `MPItrampoline_jll` package before I can properly test it?

**Comment by giordano** [comment]
No, but you need to tweak the dependency: https://docs.binarybuilder.org/stable/FAQ/#Can-I-publish-a-JLL-package-locally-without-going-through-Yggdrasil?-1.  It can be anything accepted by [`PackageSpec`](https://pkgdocs.julialang.org/v1/api/#Pkg.PackageSpec), the documentation shows an example of a remote package, but can also be a local directory.

**Comment by eschnett** [comment]
Two MPI-using packages build (`ADIOS2` and `openPMD_api`). I couldn't get the respective Julia package to work since I couldn't figure out the local deployment: Running `test ADIOS2` complains that the library `libadios2_c` cannot be found, but I confirm it was built and is thus (presumably) part of `ADIOS2_jll`.

My new plan it to ask you to merge this, and then test with the official deployment methods.

**Comment by giordano** [comment]
For the record, when you build and deploy a package (either locally, `--deploy=local`, or to a remote repository, `--deploy=username/package_name_jll.jl`) you always get a local JLL in `~/.julia/dev/package_name_jll`, so you could use as dependency
```julia
Dependency(PackageSpec(; name="package_name_jll",  uuid="...", path="/home/user/.julia/dev/package_name_jll"))
```

**Comment by eschnett** [comment]
That dependency didn't work for me. `PackageSpec` expects a git repository. I had to copy these directories and run `git init` there. (I could probably have run `git init` there locally, but I didn't want to risk messing something up.) I used e.g.
```Julia
    Dependency(PackageSpec(name="MPItrampoline_jll",
                           uuid="f1f71cc9-e9ae-5b93-9b94-4fe0e1ad3748",
                           path="/tmp/MPItrampoline_jll")),
```

This worked for a single package. I now have `ADIOS2.jl`, which requires `ADIOS2_jll`, which requires `MPItrampoline_jll`, and the libraries in `ADIOS2_jll` didn't load properly.

**Pull request merged**